geospatial page (#142)
mongodb/docs-golang: source/fundamentals/geo.txt

.. _golang-geo:

=========================
Work with Geospatial Data
=========================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to work with **geospatial data**; data formats,

> indexes, and queries. Geospatial

Author: I feel this is unnecessary due to the list of items being very short, and it would break up the flow of what you can learn and what it is.

> indexes, and queries. Geospatial

Reviewer: Resolved convo at team sync, fine to leave as is

data represents a geographic location on the surface of the Earth, or data
on a Euclidean plane.

Examples of geospatial data include:

- Locations of movie theaters
- Borders of countries
- Routes of bicycle rides
- Dog exercise areas in New York City
- Points on a graph

Store Geospatial Data
---------------------

> You have two options to store geospatial data in MongoDB:

Reviewer 2: Issue: "You have two options to store..." doesn't parse quite right.

```suggestion
All geospatial data in MongoDB is stored in one of two formats:
```

- GeoJSON, a format that represents geospatial data on an earth-like
  sphere.

- Legacy Coordinate Pair, a format that represents geospatial data
  on a Euclidean plane.

GeoJSON
~~~~~~~

GeoJSON is the preferred format to store geospatial data in MongoDB.

> Specifying a location begins with declaring one or more **positions**,
> and choosing the appropriate **type**.

Reviewer 2: Issue: Gerunds weaken writing.

```suggestion
To specify a location, declare one or more **positions** and choose the appropriate **type**.
```

https://www.mongodb.com/docs/meta/style-guide/writing/use-effective-verbs/

GeoJSON Positions

> `````````````````

Reviewer: Resolved convo at team sync. Fine to leave as is.

A position represents a single place on Earth and exists in code as an array
containing the following values:

- Longitude in the first position (required)
- Latitude in the second position (required)
- Elevation in the third position (optional)

The following is the **position** of the MongoDB Headquarters in
New York City, NY.

.. code-block:: go
   :copyable: False

   []float64{-73.986805, 40.7620853}

.. important:: Longitude then Latitude

  GeoJSON orders coordinates as **longitude** first and **latitude** second.
  This may be surprising as geographic coordinate system conventions generally list
  latitude first and longitude second. Make sure to check what format any other
  tools you are working with use. Popular tools such as OpenStreetMap and Google
  Maps list coordinates as latitude first and longitude second.

GeoJSON Types
`````````````

> Your GeoJSON object's type determines its geometric shape. Geometric shapes are
> made up of positions.

Reviewer 2: Issue: Confused about meaning of "geometric shape" here (shape of thing on grid vs. shape of JSON object).

```suggestion
Your GeoJSON object's type determines the geometric shape it represents. Geometric shapes are
made up of positions.
```

Here are some common GeoJSON types and how you can specify them with positions:

- ``Point``: a single position. This could represent the location of the
  MongoDB Headquarters.

  .. code-block:: go
     :copyable: False

>      bson.D{
>          {"name", "HQ"},
>          {"location", bson.D{
>              {"type", "Point"},
>              {"coordinates", []float64{-73.986805, 40.7620853}},
>          }},
>      }

Reviewer 2: Issue: I wasn't sure which parts of this code sample are the position vs. GeoJSON vs. other code.

Suggestion: If the "name" and "location" fields aren't critical, maybe they could be omitted. 

```suggestion
    bson.D{
             {"type", "Point"},
             {"coordinates", []float64{-73.986805, 40.7620853}},
         }},
```

Author: they are critical

>      }

Author: I'm not sure this adds value, and in later types the format is different because of the shape requirements.

>      }

Reviewer: The value this intends to add is to help further distinguish a "Point" from a "Position". AKA You use a Position to construct a Point.

The meaning of point + position tripped up Mike, who is learning this for the first time while reading it, and we agreed that helping to further clarify the difference here could be helpful.

This isn't super important, so I will not block a merge because of this.

>      }

Author: I updated the text to capture the intent of your suggestion.

- ``LineString``: an array of two or more positions that forms a series of line

>   segments. This could represent
>   `the route of the Great Wall of China <https://commons.wikimedia.org/wiki/File:GreatWallChina4.png>`__.
> 

Reviewer: Suggestion: Make link between intro sentence and GeoJSON object more explicit

Issue (Mike): "could represent" -> "represents" - [avoid subjunctive mood](https://www.mongodb.com/docs/meta/style-guide/writing/write-to-the-user/#avoid-the-subjunctive-mood)

Applies to multiple examples

```suggestion
  segments. The following GeoJSON object represents
  `the route of the Great Wall of China <https://commons.wikimedia.org/wiki/File:GreatWallChina4.png>`__:

```

Author: The style guide contains guidelines, not hard and fast rules. The main point is that a linestring is an array of two or more positions that forms a series of line segments. The subjunctive is *what* it could represent, which is anything, such as the great wall of china, broadway, the apian way, etc...

Reviewer: I think this comment is more about the fact that the example below does in fact represent a portion of the route of the great wall of china, where the text is a hypothetical (not making the link between the text and the presented snippet explicitly clear).

Maybe:

This could represent a route or path. The following GeoJSON object represents the... : 

This isn't super important, so I won't have this block a merge. Mike and I both felt the current phrasing could be more effective if updated.

  .. code-block:: go
     :copyable: False

     bson.D{
         {"name", "Great Wall of China"},
         {"location", bson.D{
             {"type", "LineString"},
             {"coordinates", [][]float64{
                 {116.572, 40.430},
                 {116.570, 40.434},
                 {116.567, 40.436},
                 {116.566, 40.441},
             }}},
         },
     }

- ``Polygon``: an array of positions in which the first and last
  position are the same and enclose some space. This could represent
  `the land within Vatican City <https://commons.wikimedia.org/wiki/File:Vatican_City_map_EN.png>`__.

  .. code-block:: go
     :copyable: False

     bson.D{
         {"name", "Vatican City"},
         {"location", bson.D{
             {"type", "Polygon"},
             {"coordinates", [][][]float64{{
                 {116.572, 40.430},
                 {116.570, 40.434},
                 {116.567, 40.436},
                 {116.572, 40.430},
             }}},
         }},
     }

> To learn more about the shapes you can use in MongoDB, see the

Reviewer: Suggestion: Inconsistent terminology "shape" vs "type". Supported by discussion with Mike in which it was unclear what the difference between "shape", "polygon", "line".

```suggestion
To learn more about the GeoJSON types you can use in MongoDB, see the
```

:manual:`GeoJSON manual entry </reference/geojson/>`.

For definitive information on GeoJSON, see the
`official IETF specification <https://datatracker.ietf.org/doc/html/rfc7946>`__.

Legacy Coordinate Pairs
~~~~~~~~~~~~~~~~~~~~~~~

> You can store geospatial data using ``x`` and ``y`` coordinates on
> a two-dimensional Euclidean plane. We refer to coordinates on a two-dimensional
> plane as "legacy coordinate pairs."

Reviewer: Suggestion: Use standard formatting for introducing a definition

Suggestion (Mike): Consider adding language to distinguish legacy coordinate pair from GeoJSON. 

```suggestion
In addition to using GeoJSON, you can store location data as a **legacy coordinate pair**.
A legacy coordinate pair is an ``x`` and ``y`` value representing a point on
a two-dimensional Euclidean plane.
```

Legacy coordinate pairs have the following structure:

.. code-block:: json

   "<field name>" : [ x, y ]

Your field should contain an array of two values in which the first represents
the ``x`` axis value and the second represents the ``y`` axis value.

For more information on legacy coordinate pairs, see the
:manual:`MongoDB server manual page on legacy coordinate pairs </geospatial-queries/#legacy-coordinate-pairs>`.

.. _golang-geospatial-indexes:

Geospatial Indexes
------------------

To enable querying for geosptial data, you must first create a supporting

> index. The index types that enable geospatial queries are:
> 
> 
> - ``2dsphere`` for GeoJSON data
> - ``2d`` for legacy coordinate pairs

Reviewer 2: Issue: [Avoid using fragments to introduce lists.](https://www.mongodb.com/docs/meta/style-guide/style/lists/)

```suggestion
index. The following index types enable geospatial queries:

- ``2dsphere`` for GeoJSON data
- ``2d`` for legacy coordinate pairs
```

- ``2dsphere`` for GeoJSON data
- ``2d`` for legacy coordinate pairs
```

When using the ``$near`` operator, you can specify the following distance operators:

- ``$minDistance``
- ``$maxDistance``

When using the ``$geoWithin`` operator, you can specify the following shape operators:

- ``$box``
- ``$polygon``
- ``$center``
- ``$centerSphere``

For more information on geospatial query operators, see the
:manual:`manual entry for geospatial queries </geospatial-queries/#geospatial-query-operators>`.

Examples
--------

The following examples use the MongoDB Atlas sample dataset. You can load sample datasets into
your database on the free tier of MongoDB Atlas by following the :atlas:`Get Started with Atlas Guide
</getting-started/#atlas-getting-started>` or you can :guides:`import the sample dataset into a local MongoDB instance
</server/import/>`.

The examples use the ``theaters`` collection in the ``sample_mflix`` database
from the sample dataset. The ``theaters`` collection contains a ``2dsphere`` index
on the ``location.geo`` field.

Query by Proximity
~~~~~~~~~~~~~~~~~~

The following example queries for documents with a ``location.geo`` field
near the MongoDB Headquarters in New York City, NY. It returns documents
from nearest to farthest.

> .. code-block:: go

Reviewer: Issue: Could the expected output from the sample data set be listed?

New York City, NY.

.. code-block:: go
   :copyable: False

   []float64{-73.986805, 40.7620853}

.. important:: Longitude then Latitude

  GeoJSON orders coordinates as **longitude** first and **latitude** second.
  This may be surprising as geographic coordinate system conventions generally list
  latitude first and longitude second. Make sure to check what format any other
  tools you are working with use. Popular tools such as OpenStreetMap and Google
  Maps list coordinates as latitude first and longitude second.

Types
`````

Your GeoJSON object's type determines the geometric shape it represents. Geometric shapes are
made up of positions.

Here are some common GeoJSON types and how you can specify them with positions:

- ``Point``: a single position. The following ``Point`` represents the location of
  the MongoDB Headquarters.

  .. code-block:: go
     :copyable: False

     bson.D{
         {"name", "MongoDB HQ"},
         {"location", bson.D{
             {"type", "Point"},
             {"coordinates", []float64{-73.986805, 40.7620853}},
         }},
     }

- ``LineString``: an array of two or more positions that forms a series of line

>   segments. The following ``LineString`` represents a segment of
>   `the Great Wall of China <https://commons.wikimedia.org/wiki/File:GreatWallChina4.png>`__.

Reviewer: Suggestion: Consider adding a sentence that explicitly states some categories of location based data this type is useful for representing.

Issue: Use colon in sentence to introduce example

```suggestion
  segments. A line string can represent a path, route, or border. The following ``LineString`` represents a segment of
  `the Great Wall of China <https://commons.wikimedia.org/wiki/File:GreatWallChina4.png>`__:
```
